### PR TITLE
XAWORKFLOW-57: Include an option to include children when publishing a ND draft

### PR DESCRIPTION
**Pull request by slauriere**
Hi all, Hi @lucaa, This pull request implements the ability to include chidren on publish. The first commit in the list performs an update of the pages and the API to xwiki-platform 12.10.4, while the next ones provide the implementation itself. Comments and reviews are welcome, thank you.

### CODE REVIEW

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 1]
```diff
@@ -1375,4 +1283,261 @@ protected String getMessage(String key, String defaultMessage, List<String> para
...
+     * the document holding the rights is <code>WebPreferences</code>, the rules are persisted to objects and saved.
+     *
+     * @param doc an XWikiDocument which holds a workflow object
+     * @param rules a list of {@link org.xwiki.security.authorization.SecurityRule}
+     * @param includeChildren whether the document's workflow scope includes children or not
+     * @throws XWikiException in case an error occurs when saving the document holding the rights
+     */
+    protected void persistAndMaybeSaveRules(XWikiDocument doc, List<ReadableSecurityRule> rules, boolean includeChildren)
+        throws XWikiException
+    {
```
Comment (lucaa) [inline xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1466]: as written above, the fact that this function is doing both save and not save is a bit tricky, we should find better, it's kinda hard to work with this as an API.

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 2]
```diff
@@ -1299,11 +1299,11 @@ protected String getMessage(String key, String defaultMessage, List<String> para
      * otherwise
      * @throws XWikiException in case an error occurs
      */
-    public void copyDocument(DocumentReference source, DocumentReference target,
+    public void copyDocument(DocumentReference draft, DocumentReference target,
```
Comment (lucaa) [inline xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1302]: I'm not sure about this change, I think the "copyDocument" (which is more of a merge logic) is also used in the other direction, from the target document to the draft, when it's about unpublishing a document, for example (to be verified!). In that case, the terminology "source" and "target" is better, as it marks a direction of the copy (merge) regardless of the roles those documents play in the workflow (published or draft).

**Comment by slauriere** [comment]
Adding here for the record an issue about the current implementation related to the "hidden" status of draft child pages: if the workflow configuration is set to "mark drafts as hidden", when a new draft child page is created after the workflow has been initialized, it should be marked as hidden. A test case has been added to `test.md` even if it's not implemented yet.

File: xwiki-workflow-publication-api/pom.xml [hunk 3]
```diff
@@ -18,24 +18,28 @@
...
   </parent>
   <artifactId>xwiki-workflow-publication-api</artifactId>
   <name>XWiki Workflow Publication API</name>
   <packaging>jar</packaging>
   <dependencies>
     <dependency>
-      <groupId>org.xwiki.platform</groupId>
-      <artifactId>xwiki-platform-oldcore</artifactId>
-      <version>${xwiki.platform.version}</version>
-      <scope>provided</scope>
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/pom.xml:38]: I see, thanks for the review and the explanation, the latest commit restores this dependency.

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 4]
```diff
@@ -513,9 +558,22 @@ private void setupDraftAccess(XWikiDocument document, BaseObject workflow, XWiki
...
+    @Override
+    public boolean startWorkflow(DocumentReference docName, boolean includeChildren, String workflowConfig,
+        DocumentReference target, XWikiContext xcontext) throws XWikiException
     {
         XWikiDocument doc = xcontext.getWiki().getDocument(docName, xcontext);
 
+        // If the workflow scope includes children, check that the target is not terminal.
+        if (includeChildren && isTerminal(docName)) {
+            // TODO: put this error in context
+            return false;
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java]: Indeed, the latest commit removes this check. I checked that local rights will be set as expected for a terminal page even if its workflow has `includeChildren` set to `true` (method `persistAndMaybeSaveRules`).

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 5]
```diff
@@ -1375,4 +1283,261 @@ protected String getMessage(String key, String defaultMessage, List<String> para
...
+     * @param source a reference to a document to be copied/merged
+     * @param target a reference to the document to be created or updated
+     * @param workflowDocumentReference a reference to the workflow document holding the workflow object in the
+     * context of which the copy occurs
+     * @param publisher a reference to the user performing the action
+     * @param includeChildren <code>true</code> if the children of the source should be copied as well, <code>false</code>
+     * otherwise
+     * @throws XWikiException in case an error occurs
+     */
+    public void copyDocument(DocumentReference source, DocumentReference target,
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1296]: I agree with you. I think I kept the name from the previous implementation. Let's find a better name.

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 6]
```diff
@@ -1159,6 +1176,8 @@ protected void cleanUpIrrelevantDataFromDoc(XWikiDocument document, XWikiContext
             document.getDocumentReference()));
         document.removeXObjects(explicitReferenceDocRefResolver.resolve(RIGHTS_CLASS,
             document.getDocumentReference()));
+        document.removeXObjects(explicitReferenceDocRefResolver.resolve(GLOBAL_RIGHTS_CLASS,
+            document.getDocumentReference()));
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1174]: Indeed, certainly worth introducing a parameter for removing rights only from the top level workflow document.

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 7]
```diff
@@ -1375,4 +1283,261 @@ protected String getMessage(String key, String defaultMessage, List<String> para
...
+
+            boolean isWorkflowDocument = target.equals(workflowDocumentReference);
+            if (isWorkflowDocument && locale.equals(sourceDocument.getDefaultLocale())) {
+                // setup the workflow and target flag, if a workflow doesn't exist already - only needs to be done for default locale
+                BaseObject newWorkflow = targetDocument.getXObject(PUBLICATION_WORKFLOW_CLASS);
+                if (newWorkflow == null) {
+                    BaseObject sourceWorkflow = sourceDocument.getXObject(PUBLICATION_WORKFLOW_CLASS);
+                    newWorkflow = targetDocument.newXObject(PUBLICATION_WORKFLOW_CLASS, xcontext);
+                    newWorkflow.set(WF_STATUS_FIELDNAME, STATUS_PUBLISHED, xcontext);
+                    newWorkflow.set(WF_INCLUDE_CHILDREN_FIELDNAME, includeChildren ? 1 : 0, xcontext);
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java]: done in latest commit, thx

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java (hunk 2 above)
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java]: The unpublish method actually uses only `copyContentsToNewVersion` but all other `copy` methods use more the source and target terminology so I aligned the argument name along this line indeed.

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 8]
```diff
@@ -1375,4 +1283,261 @@ protected String getMessage(String key, String defaultMessage, List<String> para
...
+            // save the published document prepared like this
+            String defaultMessage = "Published new version of the document by {0}.";
+            try {
+                xcontext.setLocale(translatedNewDocument.getRealLocale());
+                String message = getMessage("workflow.save.publishNew", defaultMessage,
+                    Arrays.asList(stringSerializer.serialize(publisher)));
+                // setup the context to let events know that they are in the publishing context
+                xcontext.put(CONTEXTKEY_PUBLISHING, true);
+                if (!isWorkflowDocument) {
+                    observationManager.notify(new DocumentChildPublishingEvent(target, workflowDocumentReference),
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1344]: Reported as [XAWORKFLOW-77](https://jira.xwiki.org/browse/XAWORKFLOW-77) for future improvements

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1174]: Reported as https://jira.xwiki.org/browse/XAWORKFLOW-78 for future improvements

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java (hunk 1 above)
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:1461]: I reported this improvement as [XAWORKFLOW-80](https://jira.xwiki.org/browse/XAWORKFLOW-80) for the record. Maybe we can find a better a title.

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 9]
```diff
@@ -492,17 +532,22 @@ private void setupDraftAccess(XWikiDocument document, BaseObject workflow, XWiki
...
+                document.getDocumentReference()), null, Arrays.asList(Right.VIEW), RuleState.ALLOW));
+            rules.add(rightsWriter.createRule(toDocumentReferenceList(Arrays.asList(commenters),
+                document.getDocumentReference()),null, Arrays.asList(Right.VIEW, Right.COMMENT), RuleState.ALLOW));
+            persistAndMaybeSaveRules(document, rules, workflow.getIntValue(WF_INCLUDE_CHILDREN_FIELDNAME) == 1);
 
             if (wfConfig.getIntValue(WF_CONFIG_CLASS_HIDEDRAFT_FIELDNAME, 1) == 1) {
                 document.setHidden(true);
+                if (workflow.getIntValue(WF_INCLUDE_CHILDREN_FIELDNAME) == 1) {
+                    // TODO: provide contextual message
+                    updateChildrenHiddenStatus(document.getDocumentReference(), true, null);
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:549]: Reported as XAWORKFLOW-80 for future improvement. Thanks Anca

File: xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java [hunk 10]
```diff
@@ -492,17 +532,22 @@ private void setupDraftAccess(XWikiDocument document, BaseObject workflow, XWiki
...
-            removeRestOfRights(document, 3, xcontext);
+            List<ReadableSecurityRule> rules = new ArrayList<>();
+            rules.add(rightsWriter.createRule(toDocumentReferenceList(Arrays.asList(contributors, moderators,
+                validators), document.getDocumentReference()),null, Arrays.asList(Right.VIEW, Right.COMMENT,
+                Right.EDIT), RuleState.ALLOW));
+            rules.add(rightsWriter.createRule(toDocumentReferenceList(Arrays.asList(viewers),
+                document.getDocumentReference()), null, Arrays.asList(Right.VIEW), RuleState.ALLOW));
+            rules.add(rightsWriter.createRule(toDocumentReferenceList(Arrays.asList(commenters),
+                document.getDocumentReference()),null, Arrays.asList(Right.VIEW, Right.COMMENT), RuleState.ALLOW));
+            persistAndMaybeSaveRules(document, rules, workflow.getIntValue(WF_INCLUDE_CHILDREN_FIELDNAME) == 1);
```
[the remark this answers is not in the document]
Comment (slauriere) [reply xwiki-workflow-publication-api/src/main/java/org/xwiki/workflowpublication/internal/DefaultPublicationWorkflow.java:543]: Reported as XAWORKFLOW-80

**Pull request merged**